---
title: "Arsenic - 01: Import & Normalization"
author: "Jan Waelchli"
geometry: margin=2cm
output:
  pdf_document:
    toc: yes
    toc_depth: 3
---

\pagebreak

# Experimental Setup

WT maize plants and bx1 mutants with a W22 background were grown. Half of them were treated with 0 mg/kg As (CTRL) and the other half with 100 mg/kg As (arsenic). We analyse the shift of the bacterial and fungal communities in the plant rhizosphere.

```{r setup, include=FALSE, echo=F, message=F, warning=F}

##clear the objects from memory
rm(list=ls())

#knitr settings
knitr::opts_chunk$set(echo=TRUE, fig.align="center")
options(tinytex.verbose = TRUE)

#set seed
set.seed(100)

#set paths
paths <- list(DESIGN = c("design_stats.xlsx"),
              bASV_table = c("../4_output/bacteria_ASV/bacteria_DAT100.tab"),
              btaxa = c("../4_output/bacteria_ASV/bacteria_TAXA100.tab"),
              btrack = c("../4_output/bacteria_ASV/quality_check/bacteria_track.xlsx"),
              fASV_table = c("../4_output/fungi_ASV/fungi_DAT100.tab"),
              ftaxa = c("../4_output/fungi_ASV/fungi_TAXA100.tab"),
              ftrack = c("../4_output/fungi_ASV/quality_check/fungi_track.xlsx"),
              functions="functions/"
          )

## set source to file location
if (!require(rstudioapi)){install.packages("rstudioapi")}; library(rstudioapi)
setwd(dirname(getActiveDocumentContext()$path))

## load functions
source(paste0(paths$functions,"functions.R"))
functions(path=paths$functions)

## installs (if necessary) and loads libraries
libraries()

```

```{r import design, echo=F, message=F, warning=F}

#import design file
DESIGN <- read_excel(paths$DESIGN)

#split controls from DESIGN file
DESIGN_controls <- DESIGN[grepl("control", DESIGN$sample_type),]
DESIGN <- DESIGN[!(DESIGN$sample_ID %in% DESIGN_controls$sample_ID),]

#arsenic
DESIGN$arsenic[DESIGN$arsenic=="0ppm"] <- "CTRL"
DESIGN$arsenic[DESIGN$arsenic=="100ppm"] <- "arsenic"

#factors
DESIGN$genotype <- factor(DESIGN$genotype, levels=c("WT", "bx1"))
DESIGN$arsenic <- factor(DESIGN$arsenic, levels=c("CTRL", "arsenic"))

#colors
DESIGN$cols_genotype <- NA
DESIGN$cols_genotype[DESIGN$genotype=="WT"] <- "gold2"
DESIGN$cols_genotype[DESIGN$genotype=="bx1"] <- "palegreen2"

DESIGN$cols_arsenic <- NA
DESIGN$cols_arsenic[DESIGN$arsenic=="CTRL"] <- "lemonchiffon3"
DESIGN$cols_arsenic[DESIGN$arsenic=="arsenic"] <- "lemonchiffon4"

## collapsed color vectors
temp <- data.frame(DESIGN$genotype, DESIGN$cols_genotype)
temp <- plyr::ddply(temp, .variables="DESIGN.cols_genotype", .fun=unique)
level_cols_genotype <- as.character(temp[,2])
names(level_cols_genotype) <- temp[,1]

temp <- data.frame(DESIGN$arsenic, DESIGN$cols_arsenic)
temp <- plyr::ddply(temp, .variables="DESIGN.cols_arsenic", .fun=unique)
level_cols_arsenic <- as.character(temp[,2])
names(level_cols_arsenic) <- temp[,1]

```

```{r DAT import, echo=F, message=F, warning=F, include=F}

#BACTERIA

#import ASV table
all_bDAT <- read.delim(paths$bASV_table, header=T, row.names=1, sep="\t")
all_bDAT <- t(all_bDAT)

# simplify names
colnames(all_bDAT) <- gsub("_F.fastq.gz_F_filt.fastq", "", colnames(all_bDAT))
colnames(all_bDAT) <- gsub("runBE12_", "", colnames(all_bDAT))

#sort
order <- match(DESIGN$bac_FLD, colnames(all_bDAT))
bDAT <- as.data.frame(all_bDAT[, order])

#split make control bDAT
order <- match(DESIGN_controls$bac_FLD, colnames(all_bDAT))
bDAT_controls <- as.data.frame(all_bDAT[, order])

#remove 
rm(all_bDAT)


#FUNGI

#import ASV table
all_fDAT <- read.delim(paths$fASV_table, header=T, row.names=1, sep="\t")
all_fDAT <- t(all_fDAT)

# simplify names
colnames(all_fDAT) <- gsub("_F.fastq.gz_F_filt.fastq", "", colnames(all_fDAT))
colnames(all_fDAT) <- gsub("runBE12_", "", colnames(all_fDAT))


#remove samples not in fDAT
#samples_to_keep <- which(DESIGN$fun_FLD %in% colnames(all_fDAT))
#DESIGN <- DESIGN[samples_to_keep,]

#sort
order <- match(DESIGN$fun_FLD, colnames(all_fDAT))
fDAT <- as.data.frame(all_fDAT[, order])

#split make control fDAT
order <- match(DESIGN_controls$fun_FLD, colnames(all_fDAT))
fDAT_controls <- as.data.frame(all_fDAT[, order])

#remove 
rm(all_fDAT)

```

```{r bTAX import, echo=F, message=F, warning=F, include=F}

#BACTERIA

#import taxonomy table
bTAX <- read.table(paths$btaxa, row.names=1, sep="\t", blank.lines.skip = FALSE)

#rename
colnames(bTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus")
rownames(bTAX) <- gsub(">ASV","ASV", rownames(bTAX))
bTAX[is.na(bTAX)] <- "unassigned"

# define ASVs for removal
r1 <- -which(bTAX$kingdom=="Eukaryota")
r2 <- -which(bTAX$phylum=="Cyanobacteria")
r3 <- -which(bTAX$family=="Mitochondria")
r4 <- -which(is.na(bTAX)) #rows all na
ASVs_to_remove <- c(r1,r2,r3,r4)
if(length(ASVs_to_remove)>0){
  bTAX <- bTAX[ASVs_to_remove ,]
  bDAT <- bDAT[rownames(bTAX),]
}

#add ASV_ID to bTAXonomy file
bTAX$ASV_ID <- rownames(bTAX)

#levels
bTAX$phylum <- as.factor(bTAX$phylum)
levels(bTAX$phylum)[levels(bTAX$phylum) < 2 ] <- "unassigned"
# defining ASV colors by phylum (using the bTAXonomy file)
bTAX$labels <- as.character(bTAX$phylum)
# create separate bTAXonomy label specifying classes of Proteobacteria
bTAX$class <- as.factor(bTAX$class)
try(bTAX[ bTAX$class=="Alphaproteobacteria", ]$labels <- "Alphaproteobacteria")
try(bTAX[ bTAX$class=="Betaproteobacteria", ]$labels <- "Betaproteobacteria")
try(bTAX[ bTAX$class=="Gammaproteobacteria", ]$labels <- "Gammaproteobacteria")
try(bTAX[ bTAX$class=="Deltaproteobacteria", ]$labels <- "Deltaproteobacteria")
bTAX$labels <- as.factor(bTAX$labels)
# vector of colors for abundant phyla (and classes for Proteobacteria)
# will be used for graphs later
bTAX$cols_phyla <- as.character(bTAX$labels)
bTAX$cols_phyla <- "lightgrey"
try(bTAX[ bTAX$labels=="Alphaproteobacteria" , ]$cols_phyla <- "palegreen1")
try(bTAX[ bTAX$labels=="Betaproteobacteria" , ]$cols_phyla <- "palegreen3")
try(bTAX[ bTAX$labels=="Gammaproteobacteria" , ]$cols_phyla <- "palegreen4")
try(bTAX[ bTAX$labels=="Deltaproteobacteria" , ]$cols_phyla <- "olivedrab1")
try(bTAX[ bTAX$labels=="Actinobacteria" , ]$cols_phyla <- "indianred2")
try(bTAX[ bTAX$labels=="Bacteroidetes" , ]$cols_phyla <- "steelblue1")
try(bTAX[ bTAX$labels=="Firmicutes" , ]$cols_phyla <- "tan1")
try(bTAX[ bTAX$labels=="Acidobacteria" , ]$cols_phyla <- "lightsalmon4")
try(bTAX[ bTAX$labels=="Chloroflexi" , ]$cols_phyla <- "gold1")
try(bTAX[ bTAX$labels=="Verrucomicrobia", ]$cols_phyla <- "orchid3")
try(bTAX[ bTAX$labels=="Gemmatimonadetes", ]$cols_phyla <- "peachpuff3")
try(bTAX[ bTAX$labels=="Nanoarchaeaeota" , ]$cols_phyla <- "dodgerblue2")
try(bTAX[ bTAX$labels=="Planctomycetes" , ]$cols_phyla <- "pink")
try(bTAX[ bTAX$labels=="Thaumarchaeota" , ]$cols_phyla <- "goldenrod2")
try(bTAX[ bTAX$labels=="Patescibacteria" , ]$cols_phyla <- "darkgoldenrod3")
try(bTAX[ bTAX$labels=="Rokubacteria" , ]$cols_phyla <- "darkorchid3")
## collapsed color vector for each level
temp <- data.frame(bTAX$labels, bTAX$cols_phyla)
temp <- plyr::ddply(temp, .variables="bTAX.labels", .fun=unique)
bTAX_level_cols_phyla <- as.character(temp[,2])
names(bTAX_level_cols_phyla) <- temp[,1]

# remove no longer used files
rm(temp)

```

```{r fTAX import, warning=F, echo=F,  message=F, include=F}

#FUNGI

#import taxonomy tables
fTAX <- read.table(paths$ftaxa, row.names=1, sep="\t", blank.lines.skip = FALSE)

#rename fTAX
colnames(fTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus", "species")
fTAX$kingdom <- gsub("k__","", fTAX$kingdom )
fTAX$phylum <- gsub("p__","", fTAX$phylum )
fTAX$class <- gsub("c__","", fTAX$class )
fTAX$order <- gsub("o__","", fTAX$order )
fTAX$family <- gsub("f__","", fTAX$family )
fTAX$genus <- gsub("g__","", fTAX$genus )
fTAX$species <- gsub("s__","", fTAX$species )

#rename ASVs to fASVs for fungi ASVs
rownames(fTAX) <- gsub(">ASV","fASV", rownames(fTAX))


r1 <- -which(fTAX$kingdom=="Protista")
r2 <- -which(fTAX$kingdom=="Plantae")
r3 <- -which(fTAX$kingdom=="Protozoa")
r4 <- -which(fTAX$kingdom=="Animalia")
r5 <- -which(is.na(fTAX)) #rows all na
ASVs_to_remove <- c(r1,r2,r3,r4, r5)

  if(length(ASVs_to_remove)>0){
 fTAX <- fTAX[ASVs_to_remove ,]
}

#rename unassigned ASVs
fTAX[is.na(fTAX)] <- "unassigned"

#add ASV_ID to taxonomy file
fTAX$ASV_ID <- rownames(fTAX)

#levels
fTAX$phylum <- as.factor(fTAX$phylum)
levels(fTAX$phylum)[levels(fTAX$phylum) < 2 ] <- "unassigned"

#Defining fASV colors by phylum (using the taxonomy file)
fTAX$labels <- as.character(fTAX$phylum)
fTAX$labels <- as.factor(fTAX$labels)
levels(fTAX$labels)

#vector of colors for abundant phyla 
fTAX$cols_phyla <- fTAX$labels
fTAX$cols_phyla <- "lightgrey"
try(fTAX[fTAX$labels=="Ascomycota", ]$cols_phyla <- "dodgerblue2")
try(fTAX[fTAX$labels=="Basidiomycota", ]$cols_phyla <- "firebrick1")
try(fTAX[fTAX$labels=="Glomeromycota", ]$cols_phyla <- "seagreen4")
try(fTAX[fTAX$labels=="Olpidiomycota", ]$cols_phyla <- "goldenrod2")
try(fTAX[fTAX$labels=="Mortierellomycota" , ]$cols_phyla <- "mediumorchid1")
try(fTAX[fTAX$labels=="unassigned", ]$cols_phyla <- "dimgrey")

#collapsed color vector for each level
temp <- data.frame(fTAX$labels, fTAX$cols_phyla)
temp <- plyr::ddply(temp, .variables="fTAX.labels", .fun=unique)
fTAX_level_cols_phyla <- as.character(temp[,2])
names(fTAX_level_cols_phyla) <- temp[,1]

# remove no longer used files
rm(temp)

```

\vspace{5mm}

# Description all data

## Sequencing Depth

### Figure 11 \| Reads tracking

We plot the amount of reads during each pipeline step. This allows us to see where we loses reads and if the samples from the different groups behave similar.

\vspace{5mm}

```{r track sequences, eval=T, echo=F, message=F, warning=F, , fig.height=4, fig.width=10}

#BACTERIA

#import
track <- read_xlsx(paths$btrack)
colnames(track)[1] <- "FLD"
track$FLD <- gsub("runBE12_","",track$FLD)
track$FLD <- gsub("_F","",track$FLD)

#sort in order of Design
order <- match(DESIGN$bac_FLD, track$FLD)
track <- track[order,]

#add seqs after removing eukaryota seqs
track$noneukar <- colSums(bDAT)

#add genotype, arsenic
track$genotype <- as.factor(DESIGN$genotype)
track$arsenic <- as.factor(DESIGN$arsenic)
track_long <- tidyr::pivot_longer(track, cols= !(c("FLD", "genotype", "arsenic")), names_to = "steps", values_to = "reads")

track_long$genotype <- as.factor(track_long$genotype)
track_long$steps <- factor(track_long$steps, levels=c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim", "noneukar"))
levels(track_long$steps)[1] <- "demultiplexed"

track_bac <- ggplot(track_long, aes(steps, reads, col=genotype, lty=arsenic, group=paste(genotype, arsenic, sep="_")))+
              geom_jitter(alpha=0.1, width = 0.2)+
              geom_smooth(aes(), se=F, method="loess", formula="y~x")+
              theme_bw()+
              scale_color_manual(values = level_cols_genotype)+
              facet_grid(~"Number of reads")+
              ggtitle("Bacteria: Reads tracking")

#FUNGI

#import
track <- read_xlsx(paths$ftrack)
colnames(track)[1] <- "FLD"
track$FLD <- gsub("runBE12_","",track$FLD)
track$FLD <- gsub("_F","",track$FLD)

#sort in order of Design
order <- match(DESIGN$fun_FLD, track$FLD)
track <- track[order,]

#add seqs after removing eukaryota seqs
track$noneukar <- colSums(fDAT)

#add genotype groups
track$genotype <- as.factor(DESIGN$genotype)
track$arsenic <- as.factor(DESIGN$arsenic)
track_long <- tidyr::pivot_longer(track, cols= !(c("FLD", "genotype", "arsenic")), names_to = "steps", values_to = "reads")

track_long$genotype <- as.factor(track_long$genotype)
track_long$steps <- factor(track_long$steps, levels=c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim", "noneukar"))
levels(track_long$steps)[1] <- "demultiplexed"

track_fun <- ggplot(track_long, aes(steps, reads, col=genotype, lty=arsenic, group=paste(genotype, arsenic, sep="_")))+
                geom_jitter(alpha=0.1, width = 0.2)+
                geom_smooth(aes(), se=F, method="loess", formula="y~x")+
                theme_bw()+
                scale_color_manual(values = level_cols_genotype)+
                facet_grid(~"Number of reads")+
                ggtitle("Fungi: Reads tracking")


#combine the two plots (library cowplot)
plot <- cowplot::plot_grid(track_bac + theme(legend.position="none"),
               track_fun + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               nrow=1)

legend <- cowplot::get_legend(track_bac + theme(legend.position="bottom") + labs(fill="Genotype"))
fig1 <- cowplot::plot_grid(plot, legend, nrow=2, rel_heights=c(1, .1))

#print
# png("figures/fig1.png",6400, 4000, res = 600)
fig1
# dev.off()



```

\vspace{5mm}

**Conclusion:** We lose the expected amount of reads. Samples from different group behave very similar.

\pagebreak

### Number of sequences

We show the sum, range and median of sequecnes over all samples.

```{r seq numbers bacteria, echo=F, message=F, warning=F}

df <- data.frame(Taxa=c("Bacteria", "Fungi"), removed_samples=NA, 
                 sum=NA, min=NA, max=NA, median=NA)


#BACTERIA

#show sorted seq numbers
#sort(colSums(bDAT),decreasing = T)

#remove samples with very low seq numbers
threshold <-  10000
samples_to_remove <- names(which(colSums(bDAT) < threshold))

#print(paste(length(samples_to_remove_sand),"samples removed"))
df$removed_samples[df$Taxa == "Bacteria"] <- length(samples_to_remove)
#pander(sort(colSums(DAT_sand[,colnames(DAT_sand) %in% samples_to_remove]),decreasing=T))

bDESIGN <- DESIGN[!(DESIGN$bac_FLD %in% samples_to_remove),]
# bDAT <- bDAT[,!(colnames(bDAT) %in% samples_to_remove)]
# ASVs_to_remove <- names(which(rowSums(bDAT) == 0))
# bDAT <- bDAT[!(rownames(bDAT) %in% ASVs_to_remove),]
# bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

#seq numbers
df$sum[df$Taxa == "Bacteria"] <- sum(colSums(bDAT))
df$min[df$Taxa == "Bacteria"] <- min(colSums(bDAT))
df$max[df$Taxa == "Bacteria"] <- max(colSums(bDAT))
df$median[df$Taxa == "Bacteria"] <- median(colSums(bDAT))

#FUNGI

#show sorted seq numbers
#sort(colSums(fDAT),decreasing = T)

#remove samples with very low seq numbers
threshold <-  1000
samples_to_remove <- names(which(colSums(fDAT) < threshold))

#print(paste(length(samples_to_remove_sand),"samples removed"))
df$removed_samples[df$Taxa == "Fungi"] <- length(samples_to_remove)
#pander(sort(colSums(fDAT[,colnames(fDAT) %in% samples_to_remove]),decreasing=T))

fDESIGN <- DESIGN[!(DESIGN$fun_FLD %in% samples_to_remove),]
fDAT <- fDAT[,!(colnames(fDAT) %in% samples_to_remove)]
ASVs_to_remove <- names(which(rowSums(fDAT) == 0))
fDAT <- fDAT[!(rownames(fDAT) %in% ASVs_to_remove),]
fTAX <- fTAX[!(rownames(fTAX) %in% ASVs_to_remove),]

#seq numbers
df$sum[df$Taxa == "Fungi"] <- sum(colSums(fDAT))
df$min[df$Taxa == "Fungi"] <- min(colSums(fDAT))
df$max[df$Taxa == "Fungi"] <- max(colSums(fDAT))
df$median[df$Taxa == "Fungi"] <- median(colSums(fDAT))

#out
pander(df, caption="Number of sequences")

```
\vspace{5mm}

### Figure 12 \| Sequencing depth

\vspace{5mm}

```{r Figure 2 bacteria, echo=F, warning=F, message=F}

## boxplot
df <- data.frame(col_sum=colSums(bDAT),colnames=colnames(bDAT), arsenic=bDESIGN$arsenic, 
                 genotype=bDESIGN$genotype)

ylim2=boxplot.stats(df$col_sum)$stats[c(1, 5)]

seq_nr <- ggplot(data=df, aes(x=genotype, y=col_sum, fill=genotype)) + 
              geom_boxplot(position=position_dodge2(width=0.75, preserve="single"),outlier.colour = NA) + 
              geom_jitter(size=1, position=position_jitterdodge(jitter.width=0.2,dodge.width = 0.75))+
              theme_bw() +
              theme(legend.position="none", axis.text.x=element_text(size=5)) +
              ylab("seq") +
              scale_fill_manual(values=level_cols_genotype) +
              coord_cartesian(ylim=ylim2*1.05) +
              facet_grid(.~ arsenic, scales='free_x', space="free")+
              ggtitle("Bacteria: Sequencing depth")

## barplot
df <- df[order(df$col_sum, decreasing = T),]
df <- df[order(df$arsenic),]
df$x <- c(1:sum(df$arsenic == "CTRL"), 1:sum(df$arsenic == "arsenic"))

samples <-  ggplot(df, aes(x=x, y=col_sum, fill=genotype, color=genotype))+
                geom_bar(aes(width=1), size=0.1, stat = "identity", col="black", lty=0.5)+
                scale_fill_manual(values=level_cols_genotype) +
                xlab("samples")+
                ylab("seq / samples")+
                facet_grid(.~ arsenic, scales='free_x', space="free")+
                #geom_text(aes(label=colnames), angle=90, hjust=3, size=2)+
                theme_bw() +
                theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
                      strip.background = element_blank(), strip.text.x = element_blank(), legend.position = "none")

#combine the two plots (library cowplot)
bac_plot <- cowplot::plot_grid(seq_nr + theme(legend.position="none"),
               samples + theme(legend.position="none") + labs(colour=""),
               align = "hv",
               rel_heights = c(9,5),
               nrow=2)

```

```{r Figure 2 fungi, echo=F, warning=F, message=F}

## boxplot
df <- data.frame(col_sum=colSums(fDAT),colnames=colnames(fDAT), arsenic=fDESIGN$arsenic, 
                 genotype=fDESIGN$genotype)

ylim2=boxplot.stats(df$col_sum)$stats[c(1, 5)]

seq_nr <- ggplot(data=df, aes(x=genotype, y=col_sum, fill=genotype)) + 
              geom_boxplot(position=position_dodge2(width=0.75, preserve="single"),outlier.colour = NA) + 
              geom_jitter(size=1, position=position_jitterdodge(jitter.width=0.2,dodge.width = 0.75))+
              theme_bw() +
              theme(legend.position="none", axis.text.x=element_text(size=5)) +
              ylab("seq") +
              scale_fill_manual(values=level_cols_genotype) +
              coord_cartesian(ylim=ylim2*1.05) +
              facet_grid(.~ arsenic, scales='free_x', space="free")+
              ggtitle("Fungi: Sequencing depth")

## barplot
df <- df[order(df$col_sum, decreasing = T),]
df <- df[order(df$arsenic),]
df$x <- c(1:sum(df$arsenic == "CTRL"), 1:sum(df$arsenic == "arsenic"))

samples <-  ggplot(df, aes(x=x, y=col_sum, fill=genotype, color=genotype))+
                geom_bar(aes(width=1), size=0.1, stat = "identity", col="black", lty=0.5)+
                scale_fill_manual(values=level_cols_genotype) +
                xlab("samples")+
                ylab("seq / samples")+
                facet_grid(.~ arsenic, scales='free_x', space="free")+
                theme_bw() +
                #geom_text(aes(label=colnames), angle=90, hjust=3, size=2)+
                theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
                      strip.background = element_blank(), strip.text.x = element_blank(), legend.position = "none")


#combine the two plots (library cowplot)
fun_plot <- cowplot::plot_grid(seq_nr + theme(legend.position="none"),
               samples + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               rel_heights = c(9,5),
               nrow=2)

```

```{r Figure 2.1 combined, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

#combine the two plots (library cowplot)
fig2.1 <- cowplot::plot_grid(bac_plot + theme(legend.position="none"),
               fun_plot + theme(legend.position="none") + labs(colour=""),
               align = "hv",
               nrow=1)

#add legend
legend <- cowplot::get_legend(seq_nr + theme(legend.position="bottom") + labs(fill="genotype"))
fig2.1 <- cowplot::plot_grid(fig2.1, legend, nrow=2, rel_heights=c(1, .1))

#print
# png("figures/fig3.png",6400, 4000, res = 600)
fig2.1
# dev.off()

```

\pagebreak

## Normalization

### Asymptotic Kruskal-Wallis Test & Normalization

To decide on how to normalize the data we follow the recommendation of Weiss et al. (2017, Microbiome Journal) and inspect whether there are differences in sequencing depths between the different arsenic-treatments and genotypes by using the non-parametric Kruskal-Wallis Test.

```{r seq numbers per group bacteria, echo=F, message=F, warning=F}

#BACTERIA
group <- as.factor(paste(bDESIGN$genotype, bDESIGN$arsenic, sep="_"))
sample_depth <- colSums(bDAT)
DAT_krusk <- kruskal.test(sample_depth ~ group)
print("Bacteria")
DAT_krusk

#FUNGI
group <- as.factor(paste(fDESIGN$genotype, fDESIGN$arsenic, sep="_"))
sample_depth <- colSums(fDAT)
DAT_krusk <- kruskal.test(sample_depth ~ group)
print("Fungi")
DAT_krusk

```

**Conclusion:** We don't find significant differences between the groups in bacteria or fungi. We follow the recommendation of Weiss et al. (2017) to use TSS normalization for samples with small sequencing-depth differences.

```{r DAT norm, warning=F, echo=F}

#TSS normalisation
bDAT_norm <- hilldiv::tss(bDAT)
fDAT_norm <- hilldiv::tss(fDAT)

```

\vspace{5mm}

### Outlier Detection

We use the method CLOUD developed by Montassier et al. 2018, which is a non-parametric detection test for outliers. We perform the test with Bray-Curtis distances from the normalized data for each substrate and each plastic treatment individually. We set the number of nearest neighbors to 60% of the samples size and chose an empirical outlier percentile of 0.1. We remove all outliers from our data.

```{r outlier detection, warning=F, echo=F}

#function to detect outliers
detect_outliers <- function(DESIGN, DAT_norm, Arsenic, species="bacteria"){

  #subset
  if(species=="bacteria"){samples_sub <- DESIGN$bac_FLD[DESIGN$arsenic == Arsenic]}
  if(species=="fungi"){samples_sub <- DESIGN$fun_FLD[DESIGN$arsenic == Arsenic]}
  DAT_sub <- DAT_norm[,colnames(DAT_norm) %in% samples_sub]
  bdist <- vegan::vegdist(t(DAT_sub), method="bray")
  
  #test
  k <- floor(ncol(DAT_sub) * 0.6) #number of nearest neighbors, here we chose 60% of sample size
  test <- piecewise_kn_V1(d = bdist, test.ix = colnames(DAT_sub), k = k) #CLOUD test from Montassier et al. 2018 (CLOUD function)
  outliers <- colnames(DAT_sub)[test[["pvals"]]<=0.1] #outliers (p <= 0.1)
  
  return(outliers)

}

#get outliers
bac_outliers_CTRL <- detect_outliers(DESIGN=bDESIGN, DAT_norm=bDAT_norm, Arsenic="CTRL", species="bacteria")
fun_outliers_CTRL <- detect_outliers(DESIGN=fDESIGN, DAT_norm=fDAT_norm, Arsenic="CTRL", species="fungi")
bac_outliers_arsenic <- detect_outliers(DESIGN=bDESIGN, DAT_norm=bDAT_norm, Arsenic="arsenic", species="bacteria")
fun_outliers_arsenic <- detect_outliers(DESIGN=fDESIGN, DAT_norm=fDAT_norm, Arsenic="arsenic", species="fungi")


#output
outliers <- data.frame(Species=c("Bacteria","Fungi"),
                       "Arsenic:CTRL"=c(length(bac_outliers_CTRL), length(fun_outliers_CTRL)),
                       "Arsenic:arsenic"=c(length(bac_outliers_arsenic), length(fun_outliers_arsenic)))

pander(outliers, caption = "Number of outliers")

```

```{r outlier removal, warning=F, echo=F}

#bacteria
samples_to_remove_bac <- c(bac_outliers_CTRL, bac_outliers_arsenic)

bDESIGN <- bDESIGN[!(bDESIGN$bac_FLD %in% samples_to_remove_bac),]
bDAT <- bDAT[,!(colnames(bDAT) %in% samples_to_remove_bac)]
ASVs_to_remove <- names(which(rowSums(bDAT) == 0))
bDAT <- bDAT[!(rownames(bDAT) %in% ASVs_to_remove),]
bDAT_norm <- bDAT_norm[,!(colnames(bDAT_norm) %in% samples_to_remove_bac)]
ASVs_to_remove <- names(which(rowSums(bDAT_norm) == 0))
bDAT_norm <- bDAT_norm[!(rownames(bDAT_norm) %in% ASVs_to_remove),]
bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

#fungi
samples_to_remove_fun <- c(fun_outliers_CTRL, fun_outliers_arsenic)

fDESIGN <- fDESIGN[!(fDESIGN$fun_FLD %in% samples_to_remove_fun),]
fDAT <- fDAT[,!(colnames(fDAT) %in% samples_to_remove_fun)]
ASVs_to_remove <- names(which(rowSums(fDAT) == 0))
fDAT <- fDAT[!(rownames(fDAT) %in% ASVs_to_remove),]
fDAT_norm <- fDAT_norm[,!(colnames(fDAT_norm) %in% samples_to_remove_fun)]
ASVs_to_remove <- names(which(rowSums(fDAT_norm) == 0))
fDAT_norm <- fDAT_norm[!(rownames(fDAT_norm) %in% ASVs_to_remove),]
fTAX <- fTAX[!(rownames(fTAX) %in% ASVs_to_remove),]

```

\pagebreak

## Sample Control

### Sample Size

We end up with the following number of samples per treatment for the analysis.  

```{r sample size rare, warning=F, echo=F}

#final number of samples
table_bac <- table(bDESIGN$genotype, bDESIGN$arsenic)
table_fun <- table(fDESIGN$genotype, fDESIGN$arsenic)

#out
pander(table_bac, caption = "Bacteria: Sample profile")
pander(table_fun, caption = "Fungi: Sample profile")

#write.xlsx(btable, "btable.xlsx")
#write.xlsx(ftable, "ftable.xlsx")

```

### Figure 13 \| Rarefaction plot

We plot a rarefaction plot with the remaining samples to check if the sequence depth is enough to capture the microbial diversity.

```{r rarefaction, eval=T, echo=F, message=F, warning=F, fig.height=4, fig.width=10}

#plot rarefaction

#BACTERIA

#rarefy
pdf("/dev/null") #prevents drawing of standard rarefaction plot (we will create our own with colors)
brarefaction <- vegan::rarecurve(t(bDAT), step=200, 7000, label = FALSE)
names(brarefaction) <- bDESIGN$bac_FLD
invisible(dev.off())

# long-transformation.
protox <- mapply(FUN = function(x, y) {
  mydf <- as.data.frame(x)
  colnames(mydf) <- "value"
  mydf$species <- y
  mydf$subsample <- attr(x, "Subsample")
  mydf
}, x = brarefaction, y = as.list(names(brarefaction)), SIMPLIFY = FALSE)

brarefaction_long <- do.call(rbind, protox)
rownames(brarefaction_long) <- NULL  # pretty

#add genotype
brarefaction_long$genotype <- NA
for (i in 1:nrow(brarefaction_long)) {
  brarefaction_long$genotype[i] <- as.character(DESIGN$genotype[DESIGN$bac_FLD == brarefaction_long$species[i]])
}

#add arsenic
brarefaction_long$arsenic <- NA
for (i in 1:nrow(brarefaction_long)) {
  brarefaction_long$arsenic[i] <- as.character(DESIGN$arsenic[DESIGN$bac_FLD == brarefaction_long$species[i]])
}


#plot
bac_rareplot <- ggplot(brarefaction_long, aes(x = subsample, y = value, group = species)) +
                  theme_bw() +
                  geom_line(aes(color=genotype, lty=arsenic))+
                  scale_color_manual(values = level_cols_genotype)+
                  xlab("Sequencing Depth")+
                  ylab("Number of ASVs")+
                  ggtitle("Bacteria: Rarefaction")

#FUNGI

#rarefy
pdf("/dev/null") #prevents drawing of standard rarefaction plot (we will create our own with colors)
frarefaction <- vegan::rarecurve(t(fDAT), step=10, 7000, label = FALSE)
names(frarefaction) <- fDESIGN$fun_FLD
invisible(dev.off())

# long-transformation.
protox <- mapply(FUN = function(x, y) {
  mydf <- as.data.frame(x)
  colnames(mydf) <- "value"
  mydf$species <- y
  mydf$subsample <- attr(x, "Subsample")
  mydf
}, x = frarefaction, y = as.list(names(frarefaction)), SIMPLIFY = FALSE)

frarefaction_long <- do.call(rbind, protox)
rownames(frarefaction_long) <- NULL  # pretty

#add genotype
frarefaction_long$genotype <- NA
for (i in 1:nrow(frarefaction_long)) {
  frarefaction_long$genotype[i] <- as.character(DESIGN$genotype[DESIGN$fun_FLD == frarefaction_long$species[i]])
}

#add arsenic
frarefaction_long$arsenic <- NA
for (i in 1:nrow(frarefaction_long)) {
  frarefaction_long$arsenic[i] <- as.character(DESIGN$arsenic[DESIGN$fun_FLD == frarefaction_long$species[i]])
}

#plot
fun_rareplot <- ggplot(frarefaction_long, aes(x = subsample, y = value, group = species)) +
                  theme_bw() +
                  geom_line(aes(color=genotype, lty=arsenic))+
                  scale_color_manual(values = level_cols_genotype)+
                  xlab("Sequencing Depth")+
                  ylab("Number of ASVs")+
                  ggtitle("Fungi: Rarefaction")

#combine the two plots (library cowplot)
plot <- cowplot::plot_grid(bac_rareplot + theme(legend.position="none"),
               fun_rareplot + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               nrow=1)

legend <- cowplot::get_legend(bac_rareplot + theme(legend.position="bottom") + labs(fill="Genotype"))
fig3 <- cowplot::plot_grid(plot, legend, nrow=2, rel_heights=c(1, .1))

#print
#png("figs_for_paper/fig3.png",4000, 2800, res = 600)
fig3
#dev.off()
 
```

\vspace{5mm}

**Conclusion:** All samples were sequenced deep enough.

\pagebreak

```{r export RDA files, echo=F, warning=F, message=F}

# Export

# create directory
dir.create("interim")
dir.create("interim/01_Import_Normalization")

## set output directory
setwd("interim/01_Import_Normalization")

#save objects needed in the following scripts
saveRDS(bDESIGN, "bDESIGN.RDS")
saveRDS(bDAT, "bDAT.RDS")
saveRDS(bTAX, "bTAX.RDS")
saveRDS(bDAT_norm, "bDAT_norm.RDS")

saveRDS(fDESIGN, "fDESIGN.RDS")
saveRDS(fDAT, "fDAT.RDS")
saveRDS(fTAX, "fTAX.RDS")
saveRDS(fDAT_norm, "fDAT_norm.RDS")

saveRDS(level_cols_genotype, "level_cols_genotype.RDS")
saveRDS(level_cols_arsenic, "level_cols_arsenic.RDS")
saveRDS(bTAX_level_cols_phyla, "bTAX_level_cols_phyla.RDS")
saveRDS(fTAX_level_cols_phyla, "fTAX_level_cols_phyla.RDS")

```
