---
title: "Arsenic Plant - 01: Import & Normalization"
author: "Jan Waelchli"
geometry: margin=2cm
output:
  pdf_document:
    toc: yes
    toc_depth: 3

---

\pagebreak

```{r setup, include=FALSE, echo=F, message=F, warning=F}

##clear the objects from memory
rm(list=ls())

#knitr settings
knitr::opts_chunk$set(echo=TRUE, fig.align="center")
options(tinytex.verbose = TRUE)

#set seed
set.seed(100)

#set paths
paths <- list(DESIGN = c("../../1_metafiles/design_AS_plant.xlsx"),
              bASV_table = c("../../4_output/bacteria_ASV/bacteria_ASV.tab"),
              fASV_table = c("../../4_output/fungi_ASV/fungi_ASV.tab"),
              btaxa = c("../../4_output/bacteria_ASV/bacteria_taxa.tab"),
              ftaxa = c("../../4_output/fungi_ASV/fungi_taxa.tab"),
              functions="functions/"
          )

## set source to file location
library(rstudioapi)
setwd(dirname(getActiveDocumentContext()$path))

## load functions
source(paste0(paths$functions,"functions.R"))
functions(path="functions/")

## installs (if necessary) and loads libraries
libraries()
library(dplyr)
library(tidyverse)

```

```{r import design, echo=F, message=F, warning=F}

#import design file
DESIGN <- read_excel(paths$DESIGN)

#keep only the part of the timeline experiment
DESIGN$Type <- gsub("Kernels", "Kernel", DESIGN$Type) #adapt different names
DESIGN$Type <- gsub("Kernel", "Kernels", DESIGN$Type)
DESIGN$Type <- factor(DESIGN$Type, levels = c("Control","Leaves", "Roots", "Soil", "Kernels"))
DESIGN <- droplevels(DESIGN[DESIGN$Type != "Control",])

#rename col names
colnames(DESIGN) <- gsub(" ", "_", colnames(DESIGN))

#factors
DESIGN$Sample_ID <- as.factor(gsub(" ", "", DESIGN$Sample_ID))
DESIGN$Arsenic <- factor(gsub(" ","", DESIGN$Arsenic), levels = c("0ppm","100ppm","200ppm"))
DESIGN$Soil_treatment <- factor(DESIGN$Soil_treatment, levels = c("Native"))
DESIGN$Experiment <- factor(gsub(" ", "_", DESIGN$Experiment))

#additional columns
DESIGN$group <- paste(DESIGN$Arsenic, DESIGN$Type, sep="_")
DESIGN$group <- factor(DESIGN$group, levels = c("0ppm_Leaves", "0ppm_Roots", "0ppm_Soil", "0ppm_Kernels",
                                                "100ppm_Leaves", "100ppm_Roots", "100ppm_Soil", "100ppm_Kernels",
                                                "200ppm_Leaves", "200ppm_Roots", "200ppm_Soil", "200ppm_Kernels")) 
#colors
DESIGN$cols_Arsenic <- NA
DESIGN$cols_Arsenic[DESIGN$Arsenic=="0ppm"] <- "olivedrab4"
DESIGN$cols_Arsenic[DESIGN$Arsenic=="100ppm"] <- "orange"
DESIGN$cols_Arsenic[DESIGN$Arsenic=="200ppm"] <- "orangered4"

DESIGN$cols_Type <- NA
DESIGN$cols_Type[DESIGN$Type=="Leaves"] <- "gold1"
DESIGN$cols_Type[DESIGN$Type=="Roots"] <- "orange1"
DESIGN$cols_Type[DESIGN$Type=="Soil"] <- "darkorange1"
DESIGN$cols_Type[DESIGN$Type=="Kernels"] <- "orangered1"

## collapsed color vectors
temp <- data.frame(DESIGN$Arsenic, DESIGN$cols_Arsenic)
temp <- plyr::ddply(temp, .variables="DESIGN.cols_Arsenic", .fun=unique)
level_cols_Arsenic <- as.character(temp[,2])
names(level_cols_Arsenic) <- temp[,1]

temp <- data.frame(DESIGN$Type, DESIGN$cols_Type)
temp <- plyr::ddply(temp, .variables="DESIGN.cols_Type", .fun=unique)
level_cols_Type <- as.character(temp[,2])
names(level_cols_Type) <- temp[,1]

#split to bacteria and fungi
bDESIGN <- droplevels(DESIGN[DESIGN$Library=="16S",])
fDESIGN <- droplevels(DESIGN[DESIGN$Library=="ITS F1/R2",])

```

```{r DAT import, echo=F, message=F, warning=F, include=F}

### BACTERIA ###

#import ASV table
all_bDAT <- read.delim(paths$bASV_table, header=T, row.names=1, sep="\t")
all_bDAT <- t(all_bDAT)

#keep only sample from BE08 (BE05 are soil data)
all_bDAT <- all_bDAT[,grepl("BE08", colnames(all_bDAT),fixed=T)]

## sort in order of bDESIGN and delete other cols (remove other experiments)
colnames(all_bDAT) <- gsub("runBE08_", "", colnames(all_bDAT))
colnames(all_bDAT) <- gsub("-r1_F_filt.fastq", "", colnames(all_bDAT))
colnames(all_bDAT) <- gsub("R", "_R", colnames(all_bDAT))

bDAT <- as.data.frame(all_bDAT[,colnames(all_bDAT) %in%  bDESIGN$Barcode_ID])
rm(all_bDAT)

### FUNGI ###

#import ASV table
all_fDAT <- read.delim(paths$fASV_table, header=T, row.names=1, sep="\t")
all_fDAT <- t(all_fDAT)

#keep only sample from BE08 (BE05 are soil data)
all_fDAT <- all_fDAT[,grepl("BE08", colnames(all_fDAT),fixed=T)]

## sort in order of bDESIGN and delete other cols (remove other experiments)
colnames(all_fDAT) <- gsub("runBE08_", "", colnames(all_fDAT))
colnames(all_fDAT) <- gsub("-r1_F_filt.fastq", "", colnames(all_fDAT))
colnames(all_fDAT) <- gsub("R", "_R", colnames(all_fDAT))

fDAT <- as.data.frame(all_fDAT[,colnames(all_fDAT) %in%  fDESIGN$Barcode_ID])
rm(all_fDAT)

```

```{r bTAX import, echo=F, message=F, warning=F, include=F}

### BACTERIA ###

#import taxonomy table
bTAX <- read.table(paths$btaxa, row.names=1, sep="\t", blank.lines.skip = FALSE)

#rename
colnames(bTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus")
rownames(bTAX) <- gsub(">ASV","ASV", rownames(bTAX))
bTAX[is.na(bTAX)] <- "unassigned"

# define ASVs for removal
r1 <- -which(bTAX$kingdom=="Eukaryota")
r2 <- -which(bTAX$phylum=="Cyanobacteria")
r3 <- -which(bTAX$family=="Mitochondria")
r4 <- -which(is.na(bTAX)) #rows all na
ASVs_to_remove <- c(r1,r2,r3,r4)
if(length(ASVs_to_remove)>0){
  bTAX <- bTAX[ASVs_to_remove ,]
  bDAT <- bDAT[rownames(bTAX),]
}

#add ASV_ID to bTAXonomy file
bTAX$ASV_ID <- rownames(bTAX)

#levels
bTAX$phylum <- as.factor(bTAX$phylum)
levels(bTAX$phylum)[levels(bTAX$phylum) < 2 ] <- "unassigned"
# defining ASV colors by phylum (using the bTAXonomy file)
bTAX$labels <- as.character(bTAX$phylum)
# create separate bTAXonomy label specifying classes of Proteobacteria
bTAX$class <- as.factor(bTAX$class)
try(bTAX[ bTAX$class=="Alphaproteobacteria", ]$labels <- "Alphaproteobacteria")
try(bTAX[ bTAX$class=="Betaproteobacteria", ]$labels <- "Betaproteobacteria")
try(bTAX[ bTAX$class=="Gammaproteobacteria", ]$labels <- "Gammaproteobacteria")
try(bTAX[ bTAX$class=="Deltaproteobacteria", ]$labels <- "Deltaproteobacteria")
bTAX$labels <- as.factor(bTAX$labels)
# vector of colors for abundant phyla (and classes for Proteobacteria)
# will be used for graphs later
bTAX$cols_phyla <- as.character(bTAX$labels)
bTAX$cols_phyla <- "lightgrey"
try(bTAX[ bTAX$labels=="Alphaproteobacteria" , ]$cols_phyla <- "palegreen1")
try(bTAX[ bTAX$labels=="Betaproteobacteria" , ]$cols_phyla <- "palegreen3")
try(bTAX[ bTAX$labels=="Gammaproteobacteria" , ]$cols_phyla <- "palegreen4")
try(bTAX[ bTAX$labels=="Deltaproteobacteria" , ]$cols_phyla <- "olivedrab1")
try(bTAX[ bTAX$labels=="Actinobacteria" , ]$cols_phyla <- "indianred2")
try(bTAX[ bTAX$labels=="Bacteroidetes" , ]$cols_phyla <- "steelblue1")
try(bTAX[ bTAX$labels=="Firmicutes" , ]$cols_phyla <- "tan1")
try(bTAX[ bTAX$labels=="Acidobacteria" , ]$cols_phyla <- "lightsalmon4")
try(bTAX[ bTAX$labels=="Chloroflexi" , ]$cols_phyla <- "gold1")
try(bTAX[ bTAX$labels=="Verrucomicrobia", ]$cols_phyla <- "orchid3")
try(bTAX[ bTAX$labels=="Gemmatimonadetes", ]$cols_phyla <- "peachpuff3")
try(bTAX[ bTAX$labels=="Nanoarchaeaeota" , ]$cols_phyla <- "dodgerblue2")
try(bTAX[ bTAX$labels=="Planctomycetes" , ]$cols_phyla <- "pink")
try(bTAX[ bTAX$labels=="Thaumarchaeota" , ]$cols_phyla <- "goldenrod2")
try(bTAX[ bTAX$labels=="Latescibacteria" , ]$cols_phyla <- "darkgoldenrod3")
try(bTAX[ bTAX$labels=="Rokubacteria" , ]$cols_phyla <- "darkorchid3")
## collapsed color vector for each level
temp <- data.frame(bTAX$labels, bTAX$cols_phyla)
temp <- plyr::ddply(temp, .variables="bTAX.labels", .fun=unique)
bTAX_level_cols_phyla <- as.character(temp[,2])
names(bTAX_level_cols_phyla) <- temp[,1]

#remove reconst ASVs
bTAX <- bTAX[rownames(bTAX) %in% rownames(bDAT),]

# remove no longer used files
rm(temp)

```

```{r fTAX import, echo=F, message=F, warning=F, include=F}

### FUNGI ###

#import taxonomy table
fTAX <- read.table(paths$ftaxa, row.names=1, sep="\t", blank.lines.skip = FALSE)

#rename
#colnames(fTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus")
rownames(fTAX) <- gsub(">ASV","ASV", rownames(fTAX))
fTAX[is.na(fTAX)] <- "unassigned"

#rename fTAX
colnames(fTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus", "species")
fTAX$kingdom <- gsub("k__","", fTAX$kingdom )
fTAX$phylum <- gsub("p__","", fTAX$phylum )
fTAX$class <- gsub("c__","", fTAX$class )
fTAX$order <- gsub("o__","", fTAX$order )
fTAX$family <- gsub("f__","", fTAX$family )
fTAX$genus <- gsub("g__","", fTAX$genus )
fTAX$species <- gsub("s__","", fTAX$species )

# define ASVs for removal
r1 <- -which(fTAX$kingdom=="Protista")
r2 <- -which(fTAX$kingdom=="Plantae")
r3 <- -which(fTAX$kingdom=="Protozoa")
r4 <- -which(fTAX$kingdom=="Animalia")
r5 <- -which(is.na(fTAX)) #rows all na
ASVs_to_remove <- c(r1,r2,r3,r4, r5)
if(length(ASVs_to_remove)>0){
 fTAX <- fTAX[ASVs_to_remove ,]
}

#rename unassigned ASVs
fTAX[is.na(fTAX)] <- "unassigned"

#add ASV_ID to taxonomy file
fTAX$ASV_ID <- rownames(fTAX)

#levels
fTAX$phylum <- as.factor(fTAX$phylum)
levels(fTAX$phylum)[levels(fTAX$phylum) < 2 ] <- "unassigned"
#Defining fASV colors by phylum (using the taxonomy file)
fTAX$labels <- as.character(fTAX$phylum)
fTAX$labels <- as.factor(fTAX$labels)
levels(fTAX$labels)
#vector of colors for abundant phyla 
fTAX$cols_phyla <- fTAX$labels
fTAX$cols_phyla <- "lightgrey"
try(fTAX[fTAX$labels=="Ascomycota", ]$cols_phyla <- "dodgerblue2")
try(fTAX[fTAX$labels=="Basidiomycota", ]$cols_phyla <- "firebrick1")
try(fTAX[fTAX$labels=="Olpidiomycota", ]$cols_phyla <- "seagreen4")
try(fTAX[fTAX$labels=="Chytridiomycota", ]$cols_phyla <- "goldenrod2")
try(fTAX[fTAX$labels=="Mortierellomycota" , ]$cols_phyla <- "mediumorchid1")
try(fTAX[fTAX$labels=="unassigned" , ]$cols_phyla <- "darkgray")

#collapsed color vector for each level
temp <- data.frame(fTAX$labels, fTAX$cols_phyla)
temp <- plyr::ddply(temp, .variables="fTAX.labels", .fun=unique)
fTAX_level_cols_phyla <- as.character(temp[,2])
names(fTAX_level_cols_phyla) <- temp[,1]

#remove reconst ASVs
fTAX <- fTAX[rownames(fTAX) %in% rownames(fDAT),]

# remove no longer used files
rm(temp)

```

# Description all data

To decide on how to normalize the data we followed the recommendation of Weiss et al. (2017, Microbiome Journal) and we inspected whether there are differences in sequencing depths between the different sample groups utilizing the non-parametric Kruskal-Wallis Test. 


## Sequencing depth

We show the sum, range and median over all samples we include in the analysis.

### Bacteria

```{r seq numbers bacteria, echo=F, message=F, warning=F}

### BACTERIA ###

bthreshold <-5000
samples_to_remove <- names(which(colSums(bDAT) < bthreshold))
print("samples removed (low sequencing depth)")
b_low_seq <- sort(colSums(bDAT[,colnames(bDAT) %in% samples_to_remove]),decreasing=T) %>% as.data.frame()
colnames(b_low_seq) <- "Sequences"
pander(b_low_seq, caption="Bacteria: seq depth of removed samples")

bDESIGN <- bDESIGN[!(bDESIGN$Barcode_ID %in% samples_to_remove),]
bDAT <- bDAT[,!(colnames(bDAT) %in% samples_to_remove)]
ASVs_to_remove <- names(which(rowSums(bDAT) == 0))
bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

#seq numbers
paste ("sum:", sum(colSums(bDAT)))
paste ("range:", range(colSums(bDAT)))
paste ("median:", median(colSums(bDAT)))

```

\pagebreak

### Fungi

```{r seq numbers fungi, echo=F, message=F, warning=F}

### FUNGI ###

#show sorted seq numbers
#sort(colSums(fDAT),decreasing = T)

#remove samples with very low seq numbers
fthreshold <-500
samples_to_remove <- names(which(colSums(fDAT) < fthreshold))
print("samples removed (low sequencing depth)")
f_low_seq <- sort(colSums(fDAT[,colnames(fDAT) %in% samples_to_remove]),decreasing=T) %>% as.data.frame()
colnames(f_low_seq) <- "Sequences"
pander(f_low_seq, caption="Fungi: seq depth of removed samples")

fDESIGN <- fDESIGN[!(fDESIGN$Barcode_ID %in% samples_to_remove),]
fDAT <- fDAT[,!(colnames(fDAT) %in% samples_to_remove)]
ASVs_to_remove <- names(which(rowSums(fDAT) == 0))
fTAX <- fTAX[!(rownames(fTAX) %in% ASVs_to_remove),]

#seq numbers
paste ("sum:", sum(colSums(fDAT)))
paste ("range:", range(colSums(fDAT)))
paste ("median:", median(colSums(fDAT)))

```

We excluded all bacterial samples with less than `r bthreshold` sequences and fungal samples with `r fthreshold` sequences. respectively.

\pagebreak

## Figure 1 | Sequencing depth

\vspace{5mm}

```{r Figure 1 bacteria, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

## boxplot
df <- data.frame(col_sum=colSums(bDAT),colnames=colnames(bDAT), Type=bDESIGN$Type, 
                 Arsenic=bDESIGN$Arsenic, group=bDESIGN$group, Experiment=bDESIGN$Experiment, Type=as.factor(bDESIGN$Type))

ylim2=boxplot.stats(df$col_sum)$stats[c(1, 5)]

seq_nr <- ggplot(data=df, aes(x=Arsenic, y=col_sum, fill=Arsenic)) + 
              geom_boxplot(position=position_dodge2(width=0.75, preserve="single"),outlier.colour = NA) + 
              geom_jitter(aes(group=group, col=Arsenic), width=0.2, col="gray40") + 
              theme_bw() +
              theme(legend.position="none") +
              ylab("seq") +
              scale_fill_manual(values=level_cols_Arsenic) +
              scale_color_manual(values=level_cols_Arsenic) +
              coord_cartesian(ylim=ylim2*1.05) +  
              facet_grid(.~ Type, scales='free_x', space="free")+
              theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
              ggtitle("Bacteria: Sequencing depth")

## barplot
df <- df[order(df$col_sum, decreasing = T),]
# df <- df[order(df$Type),]
# temp <- table(df$Type)
# x <- vector()
# for (i in 1:length(temp)){x <- c(x, 1:temp[i])}
# df$x <- x
df$x <- 1:length(df$col_sum)



samples <-  ggplot(df, aes(x=x, y=col_sum, fill=Arsenic, color=Arsenic))+
                geom_bar(aes(width=1), size=0.1, stat = "identity", col="black", lty=0.5)+
                #scale_y_continuous(trans = 'log2')+
                scale_fill_manual(values=level_cols_Arsenic) +
                scale_colour_manual(values = level_cols_Arsenic) +
                xlab("samples")+
                ylab("seq / samples")+
                theme_bw() +
                theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
                      strip.background = element_blank(), strip.text.x = element_blank(), legend.position = "none")
                #facet_grid(.~ Type, scales='free_x', space="free")
                #ggtitle("Sequencing depth per sample")


#combine the two plots (library cowplot)
bac_plot <- plot_grid(seq_nr + theme(legend.position="none"),
               samples + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               rel_heights = c(9,5),
               nrow=2)

```

```{r Figure 1 fungi, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

## boxplot
df <- data.frame(col_sum=colSums(fDAT),colnames=colnames(fDAT), Type=fDESIGN$Type, 
                 Arsenic=fDESIGN$Arsenic, group=fDESIGN$group, Experiment=fDESIGN$Experiment, Type=as.factor(fDESIGN$Type))
#levels(df$Type) <- c("day 0","day 2","day 5")

ylim2=boxplot.stats(df$col_sum)$stats[c(1, 5)]

seq_nr <- ggplot(data=df, aes(x=Arsenic, y=col_sum, fill=Arsenic)) + 
              geom_boxplot(position=position_dodge2(width=0.75, preserve="single"),outlier.colour = NA) + 
              geom_jitter(aes(group=group, col=Arsenic), width=0.2, col="gray40") + 
              theme_bw() +
              theme(legend.position="none") +
              ylab("seq") +
              scale_fill_manual(values=level_cols_Arsenic) +
              #scale_color_manual(values=c("gray40","gray80")) +
              coord_cartesian(ylim=ylim2*1.05) +  
              facet_grid(.~ Type, scales='free_x', space="free")+
              theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
              ggtitle("Fungi: Sequencing depth")

## barplot
df <- df[order(df$col_sum, decreasing = T),]
#df <- df[order(df$Type),]
# temp <- table(df$Type)
# x <- vector()
# for (i in 1:length(temp)){x <- c(x, 1:temp[i])}
# df$x <- x
df$x <- 1:length(df$col_sum)


samples <-  ggplot(df, aes(x=x, y=col_sum, fill=Arsenic, color=Arsenic))+
                geom_bar(aes(width=1), size=0.1, stat = "identity", col="black", lty=0.5)+
                #scale_y_continuous(trans = 'log2')+
                scale_fill_manual(values=level_cols_Arsenic) +
                scale_colour_manual(values = level_cols_Arsenic) +
                xlab("samples")+
                ylab("seq / sample")+
                theme_bw() +
                theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
                      strip.background = element_blank(), strip.text.x = element_blank(), legend.position = "none")
                #facet_grid(.~ Type, scales='free_x', space="free")
                #ggtitle("Sequencing depth per sample")


#combine the two plots (library cowplot)
fun_plot <- plot_grid(seq_nr + theme(legend.position="none"),
               samples + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               rel_heights = c(9,5),
               nrow=2)

```

```{r Figure 1 combined, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

#combine the two plots (library cowplot)
fig1 <- plot_grid(bac_plot + theme(legend.position="none"),
               fun_plot + theme(legend.position="none") + labs(colour=""),
               labels=c("A","B"),
               align = "hv",
               nrow=1)

#add legend
legend <- get_legend(seq_nr + theme(legend.position="bottom") + labs(fill="Arsenic"))
fig1 <- plot_grid(fig1, legend, nrow=2, rel_heights=c(1, .1))

#print
fig1


```

\pagebreak

# Data normalization

## Asymptotic Kruskal-Wallis Test
We test for different sequencing depths between the groups.

### Bacteria

```{r seq numbers per group bacteria, echo=F, message=F, warning=F}

bDAT_krusk <- kruskal_test(colSums(bDAT) ~ bDESIGN$group)  # library(coin)
bDAT_krusk

```

### Fungi

```{r seq numbers per group fungi, echo=F, message=F, warning=F}

fDAT_krusk <- kruskal_test(colSums(fDAT) ~ fDESIGN$group)  # library(coin)
fDAT_krusk

```

**Conclusion:** We don't find significant differences between the groups in bacteria or fungi. We follow the recommendation of Weiss et al. (2017) to use TSS normalization for samples with small sequencing-depth differences.

```{r DAT norm, warning=F, echo=F}

#TSS normalisation
bDAT_norm <- hilldiv::tss(bDAT)
fDAT_norm <- hilldiv::tss(fDAT)

```

## Outlier Detection

We use the method CLOUD developped by Montassier et al. 2018, which is a non.parametric detection test for outliers. We perform the test with Bray-Curtis distances from the normalized data. We set the number of nearest neighbors to 60% of the samples size and chose an empirical outlier percentile of 5%. We remove all outliers from our data.

### Bacteria

```{r bac outlier detection, warning=F, echo=F}

#function to detect outliers
detect_outliers <- function(DESIGN, DAT_norm){

  #subset
  samples_sub <- DESIGN$Barcode_ID
  DAT_sub <- DAT_norm[,colnames(DAT_norm) %in% samples_sub]
  bdist <- vegdist(t(DAT_sub), method="bray")
  
  #test
  k <- floor(ncol(DAT_sub) * 0.6) #number of nearest neighbors, here we chose 60% of sample size
  test <- piecewise_kn_V1(d = bdist, test.ix = colnames(DAT_sub), k = k) #CLOUD test from Montassier et al. 2018
  outliers <- colnames(DAT_sub)[test[["pvals"]]<0.05] #outliers (p < 0.0.5)
  
  return(outliers)

}

### BACTERIA 
boutliers <- detect_outliers(DESIGN=bDESIGN, DAT_norm=bDAT_norm)

bDESIGN <- bDESIGN[!(bDESIGN$Barcode_ID %in% boutliers),]
bDAT <- bDAT[,!(colnames(bDAT) %in% boutliers)]
bDAT_norm <- bDAT_norm[,!(colnames(bDAT_norm) %in% boutliers)]
ASVs_to_remove <- names(which(rowSums(bDAT_norm) == 0))
bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

print("samples removed (outliers)")
print(as.character(boutliers))

```

### Fungi

```{r fun outlier detection, warning=F, echo=F}

### FUNGI
foutliers <- detect_outliers(DESIGN=fDESIGN, DAT_norm=fDAT_norm)

fDESIGN <- fDESIGN[!(fDESIGN$Barcode_ID %in% foutliers),]
fDAT <- fDAT[,!(colnames(fDAT) %in% foutliers)]
fDAT_norm <- fDAT_norm[,!(colnames(fDAT_norm) %in% foutliers)]
ASVs_to_remove <- names(which(rowSums(fDAT_norm) == 0))
fTAX <- fTAX[!(rownames(fTAX) %in% ASVs_to_remove),]

print("samples removed (outliers)")
print(as.character(foutliers))

```

\pagebreak

## Sample Control

### Sample Size

We end up with the following number of samples per treatment for the analysis:

```{r sample size norm, warning=F, echo=F}

#final number of samples
btable <- table(bDESIGN$Type, bDESIGN$Arsenic)
pander(btable, caption="Bacteria: Sample profiles")

ftable <- table(fDESIGN$Type, fDESIGN$Arsenic)
pander(ftable, caption="Fungi: Sample profiles")

```

**Conclusion:** Because there are not enough replicates for bacterial kernel samples, we exclude them for all further analyses.

```{r rarefaction, eval=T, echo=F, message=F, warning=F, fig.height=4, fig.width=10}

#remove kernel for bac
bDESIGN <- droplevels(bDESIGN[bDESIGN$Type !="Kernels",])
bDAT <- bDAT[,colnames(bDAT) %in% bDESIGN$Barcode_ID]
bDAT_norm <- bDAT_norm[,colnames(bDAT_norm) %in% bDESIGN$Barcode_ID]
ASVs_to_remove <- names(which(rowSums(bDAT_norm) == 0))
bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

```



## Figure 2 | Rarefaction plot

We plot a rarefaction plot with the remaining samples to check if the sequence depth is enough to capture the microbial diversity.

```{r rarefaction plot, eval=T, echo=F, message=F, warning=F, fig.height=4, fig.width=10}

#plot rarefaction

#BACTERIA

#rarefy
pdf("/dev/null") #prevents drawing of standard rarefaction plot (we will create our own with colors)
brarefaction <- vegan::rarecurve(t(bDAT), step=200, 7000, label = FALSE)
names(brarefaction) <- bDESIGN$Barcode_ID
invisible(dev.off())

# long-transformation.
protox <- mapply(FUN = function(x, y) {
  mydf <- as.data.frame(x)
  colnames(mydf) <- "value"
  mydf$species <- y
  mydf$subsample <- attr(x, "Subsample")
  mydf
}, x = brarefaction, y = as.list(names(brarefaction)), SIMPLIFY = FALSE)

brarefaction_long <- do.call(rbind, protox)
rownames(brarefaction_long) <- NULL  # pretty

#add Type
brarefaction_long$Type <- NA
for (i in 1:nrow(brarefaction_long)) {
  brarefaction_long$Type[i] <- as.character(DESIGN$Type[DESIGN$Barcode_ID == brarefaction_long$species[i]])
}

#add Arsenic
brarefaction_long$Arsenic <- NA
for (i in 1:nrow(brarefaction_long)) {
  brarefaction_long$Arsenic[i] <- as.character(DESIGN$Arsenic[DESIGN$Barcode_ID == brarefaction_long$species[i]])
}


#plot
bac_rareplot <- ggplot(brarefaction_long, aes(x = subsample, y = value, group = species)) +
                  theme_bw() +
                  geom_line(aes(color=Type, lty=Arsenic))+
                  scale_color_manual(values = level_cols_Type)+
                  xlab("Sequencing Depth")+
                  ylab("Number of ASVs")+
                  ggtitle("Bacteria: Rarefaction")

#FUNGI

#rarefy
pdf("/dev/null") #prevents drawing of standard rarefaction plot (we will create our own with colors)
frarefaction <- vegan::rarecurve(t(fDAT), label = FALSE)
names(frarefaction) <- fDESIGN$Barcode_ID
invisible(dev.off())

# long-transformation.
protox <- mapply(FUN = function(x, y) {
  mydf <- as.data.frame(x)
  colnames(mydf) <- "value"
  mydf$species <- y
  mydf$subsample <- attr(x, "Subsample")
  mydf
}, x = frarefaction, y = as.list(names(frarefaction)), SIMPLIFY = FALSE)

frarefaction_long <- do.call(rbind, protox)
rownames(frarefaction_long) <- NULL  # pretty

#add Type
frarefaction_long$Type <- NA
for (i in 1:nrow(frarefaction_long)) {
  frarefaction_long$Type[i] <- as.character(DESIGN$Type[DESIGN$Barcode_ID == frarefaction_long$species[i]])
}

#add Arsenic
frarefaction_long$Arsenic <- NA
for (i in 1:nrow(frarefaction_long)) {
  frarefaction_long$Arsenic[i] <- as.character(DESIGN$Arsenic[DESIGN$Barcode_ID == frarefaction_long$species[i]])
}

#plot
fun_rareplot <- ggplot(frarefaction_long, aes(x = subsample, y = value, group = species)) +
                  theme_bw() +
                  geom_line(aes(color=Type, lty=Arsenic))+
                  scale_color_manual(values = level_cols_Type)+
                  xlab("Sequencing Depth")+
                  ylab("Number of ASVs")+
                  ggtitle("Fungi: Rarefaction")

#combine the two plots (library cowplot)
plot <- cowplot::plot_grid(bac_rareplot + theme(legend.position="none"),
               fun_rareplot + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               nrow=1)

legend <- cowplot::get_legend(bac_rareplot + theme(legend.position="bottom") + labs(fill="Genotype"))
fig3 <- cowplot::plot_grid(plot, legend, nrow=2, rel_heights=c(1, .1))

#print
#png("figs_for_paper/fig3.png",4000, 2800, res = 600)
fig3
#dev.off()
 
```

\vspace{5mm}

**Conclusion:** The remaining samples are sequenced deep enough to cover the majority of microbial diversity.

```{r export RDA files, echo=F, warning=F, message=F}

# Export

# create directory
dir.create("interim")
dir.create("interim/01_Import_Normalization")

## set output directory
setwd("interim/01_Import_Normalization")

#save objects needed in the following scripts
saveRDS(bDESIGN, "bDESIGN.RDS")
saveRDS(bDAT, "bDAT.RDS")
saveRDS(bTAX, "bTAX.RDS")
saveRDS(bDAT_norm, "bDAT_norm.RDS")

saveRDS(fDESIGN, "fDESIGN.RDS")
saveRDS(fDAT, "fDAT.RDS")
saveRDS(fTAX, "fTAX.RDS")
saveRDS(fDAT_norm, "fDAT_norm.RDS")

saveRDS(level_cols_Arsenic, "level_cols_Arsenic.RDS")
saveRDS(level_cols_Type, "level_cols_Type.RDS")
saveRDS(bTAX_level_cols_phyla, "bTAX_level_cols_phyla.RDS")
saveRDS(fTAX_level_cols_phyla, "fTAX_level_cols_phyla.RDS")

```

\pagebreak
