---
title: "Arsenic Soil - 01: Import & Normalization"
author: "Jan Waelchli"
geometry: margin=2cm
output:
  pdf_document:
    toc: yes
    toc_depth: 3

---

\pagebreak

```{r setup, include=FALSE, echo=F, message=F, warning=F}

##clear the objects from memory
rm(list=ls())

#knitr settings
knitr::opts_chunk$set(echo=TRUE, fig.align="center")
options(tinytex.verbose = TRUE)

#set seed
set.seed(100)

#set paths
paths <- list(DESIGN = c("../../1_metafiles/design_AS_soil.xlsx"),
              bASV_table = c("../../4_output/bacteria_ASV/bacteria_ASV.tab"),
              fASV_table = c("../../4_output/fungi_ASV/fungi_ASV.tab"),
              btaxa = c("../../4_output/bacteria_ASV/bacteria_taxa.tab"),
              ftaxa = c("../../4_output/fungi_ASV/fungi_taxa.tab"),
              functions="functions/"
          )

## set source to file location
library(rstudioapi)
setwd(dirname(getActiveDocumentContext()$path))

## load functions
source(paste0(paths$functions,"functions.R"))
functions(path="functions/")

## installs (if necessary) and loads libraries
libraries()
library(dplyr)
library(tidyverse)

```

```{r import design, echo=F, message=F, warning=F}

#import design file
DESIGN <- read_excel(paths$DESIGN)

# #keep only the part of the timeline experiment
# DESIGN$Experiment <- factor(DESIGN$Experiment, levels = c("timeline","reconst"))
# DESIGN <- droplevels(DESIGN[DESIGN$Experiment == "timeline",])

#factors
DESIGN$SampleType <- factor(DESIGN$SampleType, levels = c("soil"))
DESIGN$Arsenic <- factor(DESIGN$Arsenic, levels = c("0ppm","100ppm","200ppm"))
DESIGN$Timepoint <- factor(DESIGN$Timepoint, levels = c("0M","1M","2M","3M","4M","5M"))
DESIGN$Soil <- factor(DESIGN$Soil, levels = c("normal"))
DESIGN$Taxa <- factor(DESIGN$Taxa, levels = c("b","f"))

#additional columns
DESIGN$group <- paste(DESIGN$Arsenic, DESIGN$Timepoint, sep="_")
DESIGN$group <- factor(DESIGN$group, levels = c("0ppm_0M", "0ppm_1M", "0ppm_2M", "0ppm_3M", "0ppm_4M", "0ppm_5M",
                                                "100ppm_0M", "100ppm_1M", "100ppm_2M", "100ppm_3M", "100ppm_4M", "100ppm_5M",
                                                "200ppm_0M", "200ppm_1M", "200ppm_2M", "200ppm_3M", "200ppm_4M", "200ppm_5M")) 
#colors
DESIGN$cols_Arsenic <- NA
DESIGN$cols_Arsenic[DESIGN$Arsenic=="0ppm"] <- "olivedrab4"
DESIGN$cols_Arsenic[DESIGN$Arsenic=="100ppm"] <- "orange"
DESIGN$cols_Arsenic[DESIGN$Arsenic=="200ppm"] <- "orangered4"

DESIGN$cols_Timepoint <- NA
DESIGN$cols_Timepoint[DESIGN$Timepoint=="0M"] <- "darkgrey"
DESIGN$cols_Timepoint[DESIGN$Timepoint=="1M"] <- "gold1"
DESIGN$cols_Timepoint[DESIGN$Timepoint=="2M"] <- "orange1"
DESIGN$cols_Timepoint[DESIGN$Timepoint=="3M"] <- "darkorange1"
DESIGN$cols_Timepoint[DESIGN$Timepoint=="4M"] <- "orangered1"
DESIGN$cols_Timepoint[DESIGN$Timepoint=="5M"] <- "orangered4"

## collapsed color vectors
temp <- data.frame(DESIGN$Arsenic, DESIGN$cols_Arsenic)
temp <- plyr::ddply(temp, .variables="DESIGN.cols_Arsenic", .fun=unique)
level_cols_Arsenic <- as.character(temp[,2])
names(level_cols_Arsenic) <- temp[,1]

temp <- data.frame(DESIGN$Timepoint, DESIGN$cols_Timepoint)
temp <- plyr::ddply(temp, .variables="DESIGN.cols_Timepoint", .fun=unique)
level_cols_Timepoint <- as.character(temp[,2])
names(level_cols_Timepoint) <- temp[,1]

#split to bacteria and fungi
bDESIGN <- droplevels(DESIGN[DESIGN$Taxa=="b",])
fDESIGN <- droplevels(DESIGN[DESIGN$Taxa=="f",])

```

```{r DAT import, echo=F, message=F, warning=F, include=F}

### BACTERIA ###

#import ASV table
all_bDAT <- read.delim(paths$bASV_table, header=T, row.names=1, sep="\t")
all_bDAT <- t(all_bDAT)

#keep only sample from BE05 (BE08 are plant data)
all_bDAT <- all_bDAT[,grepl("BE05", colnames(all_bDAT),fixed=T)]

## sort in order of bDESIGN and delete other cols (remove other experiments)
colnames(all_bDAT) <- gsub("runBE05_", "", colnames(all_bDAT))
colnames(all_bDAT) <- gsub("-r1_F_filt.fastq", "", colnames(all_bDAT))
bDAT <- as.data.frame(all_bDAT[,colnames(all_bDAT) %in%  bDESIGN$Label])
rm(all_bDAT)

# #rm bDESIGN samples not in bDAT
# bDESIGN <- bDESIGN[bDESIGN$Label %in% colnames(bDAT),]

### FUNGI ###

#import ASV table
all_fDAT <- read.delim(paths$fASV_table, header=T, row.names=1, sep="\t")
all_fDAT <- t(all_fDAT)

## sort in order of fDESIGN and delete other cols
colnames(all_fDAT) <- gsub("runBE05_", "", colnames(all_fDAT))
colnames(all_fDAT) <- gsub("-r1_F_filt.fastq", "", colnames(all_fDAT))
fDAT <- as.data.frame(all_fDAT[,colnames(all_fDAT) %in%  fDESIGN$Label])
rm(all_fDAT)

# #rm fDESIGN samples not in fDAT
# fDESIGN <- fDESIGN[fDESIGN$Label %in% colnames(fDAT),]

```

```{r bTAX import, echo=F, message=F, warning=F, include=F}

### BACTERIA ###

#import taxonomy table
bTAX <- read.table(paths$btaxa, row.names=1, sep="\t", blank.lines.skip = FALSE)

#rename
colnames(bTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus")
rownames(bTAX) <- gsub(">ASV","ASV", rownames(bTAX))
bTAX[is.na(bTAX)] <- "unassigned"

# define ASVs for removal
r1 <- -which(bTAX$kingdom=="Eukaryota")
r2 <- -which(bTAX$phylum=="Cyanobacteria")
r3 <- -which(bTAX$family=="Mitochondria")
r4 <- -which(is.na(bTAX)) #rows all na
ASVs_to_remove <- c(r1,r2,r3,r4)
if(length(ASVs_to_remove)>0){
  bTAX <- bTAX[ASVs_to_remove ,]
  bDAT <- bDAT[rownames(bTAX),]
}

#add ASV_ID to bTAXonomy file
bTAX$ASV_ID <- rownames(bTAX)

#levels
bTAX$phylum <- as.factor(bTAX$phylum)
levels(bTAX$phylum)[levels(bTAX$phylum) < 2 ] <- "unassigned"
# defining ASV colors by phylum (using the bTAXonomy file)
bTAX$labels <- as.character(bTAX$phylum)
# create separate bTAXonomy label specifying classes of Proteobacteria
bTAX$class <- as.factor(bTAX$class)
try(bTAX[ bTAX$class=="Alphaproteobacteria", ]$labels <- "Alphaproteobacteria")
try(bTAX[ bTAX$class=="Betaproteobacteria", ]$labels <- "Betaproteobacteria")
try(bTAX[ bTAX$class=="Gammaproteobacteria", ]$labels <- "Gammaproteobacteria")
try(bTAX[ bTAX$class=="Deltaproteobacteria", ]$labels <- "Deltaproteobacteria")
bTAX$labels <- as.factor(bTAX$labels)
# vector of colors for abundant phyla (and classes for Proteobacteria)
# will be used for graphs later
bTAX$cols_phyla <- as.character(bTAX$labels)
bTAX$cols_phyla <- "lightgrey"
try(bTAX[ bTAX$labels=="Alphaproteobacteria" , ]$cols_phyla <- "palegreen1")
try(bTAX[ bTAX$labels=="Betaproteobacteria" , ]$cols_phyla <- "palegreen3")
try(bTAX[ bTAX$labels=="Gammaproteobacteria" , ]$cols_phyla <- "palegreen4")
try(bTAX[ bTAX$labels=="Deltaproteobacteria" , ]$cols_phyla <- "olivedrab1")
try(bTAX[ bTAX$labels=="Actinobacteria" , ]$cols_phyla <- "indianred2")
try(bTAX[ bTAX$labels=="Bacteroidetes" , ]$cols_phyla <- "steelblue1")
try(bTAX[ bTAX$labels=="Firmicutes" , ]$cols_phyla <- "tan1")
try(bTAX[ bTAX$labels=="Acidobacteria" , ]$cols_phyla <- "lightsalmon4")
try(bTAX[ bTAX$labels=="Chloroflexi" , ]$cols_phyla <- "gold1")
try(bTAX[ bTAX$labels=="Verrucomicrobia", ]$cols_phyla <- "orchid3")
try(bTAX[ bTAX$labels=="Gemmatimonadetes", ]$cols_phyla <- "peachpuff3")
try(bTAX[ bTAX$labels=="Nanoarchaeaeota" , ]$cols_phyla <- "dodgerblue2")
try(bTAX[ bTAX$labels=="Planctomycetes" , ]$cols_phyla <- "pink")
try(bTAX[ bTAX$labels=="Thaumarchaeota" , ]$cols_phyla <- "goldenrod2")
try(bTAX[ bTAX$labels=="Latescibacteria" , ]$cols_phyla <- "darkgoldenrod3")
try(bTAX[ bTAX$labels=="Rokubacteria" , ]$cols_phyla <- "darkorchid3")
## collapsed color vector for each level
temp <- data.frame(bTAX$labels, bTAX$cols_phyla)
temp <- plyr::ddply(temp, .variables="bTAX.labels", .fun=unique)
bTAX_level_cols_phyla <- as.character(temp[,2])
names(bTAX_level_cols_phyla) <- temp[,1]

#remove reconst ASVs
bTAX <- bTAX[rownames(bTAX) %in% rownames(bDAT),]

# remove no longer used files
rm(temp)

```

```{r fTAX import, echo=F, message=F, warning=F, include=F}

### FUNGI ###

#import taxonomy table
fTAX <- read.table(paths$ftaxa, row.names=1, sep="\t", blank.lines.skip = FALSE)

#rename
#colnames(fTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus")
rownames(fTAX) <- gsub(">ASV","ASV", rownames(fTAX))
fTAX[is.na(fTAX)] <- "unassigned"

#rename fTAX
colnames(fTAX) <- c("kingdom", "phylum", "class", "order", "family", "genus", "species")
fTAX$kingdom <- gsub("k__","", fTAX$kingdom )
fTAX$phylum <- gsub("p__","", fTAX$phylum )
fTAX$class <- gsub("c__","", fTAX$class )
fTAX$order <- gsub("o__","", fTAX$order )
fTAX$family <- gsub("f__","", fTAX$family )
fTAX$genus <- gsub("g__","", fTAX$genus )
fTAX$species <- gsub("s__","", fTAX$species )

# define ASVs for removal
r1 <- -which(fTAX$kingdom=="Protista")
r2 <- -which(fTAX$kingdom=="Plantae")
r3 <- -which(fTAX$kingdom=="Protozoa")
r4 <- -which(fTAX$kingdom=="Animalia")
r5 <- -which(is.na(fTAX)) #rows all na
ASVs_to_remove <- c(r1,r2,r3,r4, r5)
if(length(ASVs_to_remove)>0){
 fTAX <- fTAX[ASVs_to_remove ,]
}

#rename unassigned ASVs
fTAX[is.na(fTAX)] <- "unassigned"

#add ASV_ID to taxonomy file
fTAX$ASV_ID <- rownames(fTAX)

#levels
fTAX$phylum <- as.factor(fTAX$phylum)
levels(fTAX$phylum)[levels(fTAX$phylum) < 2 ] <- "unassigned"
#Defining fASV colors by phylum (using the taxonomy file)
fTAX$labels <- as.character(fTAX$phylum)
fTAX$labels <- as.factor(fTAX$labels)
levels(fTAX$labels)
#vector of colors for abundant phyla 
fTAX$cols_phyla <- fTAX$labels
fTAX$cols_phyla <- "lightgrey"
try(fTAX[fTAX$labels=="Ascomycota", ]$cols_phyla <- "dodgerblue2")
try(fTAX[fTAX$labels=="Basidiomycota", ]$cols_phyla <- "firebrick1")
try(fTAX[fTAX$labels=="Olpidiomycota", ]$cols_phyla <- "seagreen4")
try(fTAX[fTAX$labels=="Chytridiomycota", ]$cols_phyla <- "goldenrod2")
try(fTAX[fTAX$labels=="Mortierellomycota" , ]$cols_phyla <- "mediumorchid1")
try(fTAX[fTAX$labels=="unassigned" , ]$cols_phyla <- "darkgray")

#collapsed color vector for each level
temp <- data.frame(fTAX$labels, fTAX$cols_phyla)
temp <- plyr::ddply(temp, .variables="fTAX.labels", .fun=unique)
fTAX_level_cols_phyla <- as.character(temp[,2])
names(fTAX_level_cols_phyla) <- temp[,1]

#remove reconst ASVs
fTAX <- fTAX[rownames(fTAX) %in% rownames(fDAT),]

# remove no longer used files
rm(temp)

```

# Description all data

To decide on how to normalize the data we followed the recommendation of Weiss et al. (2017, Microbiome Journal) and we inspected whether there are differences in sequencing depths between the different sample groups utilizing the non-parametric Kruskal-Wallis Test. 


## Sequencing depth

We show the sum, range and median over all samples we include in the analysis.

### Bacteria

```{r seq numbers bacteria, echo=F, message=F, warning=F}

### BACTERIA ###

#show sorted seq numbers
#sort(colSums(bDAT),decreasing = T)

#remove samples with very low seq numbers
# threshold <-  1000
# samples_to_remove <- names(which(colSums(bDAT) < threshold))
# 
# print("samples removed")
# pander(sort(colSums(bDAT[,colnames(bDAT) %in% samples_to_remove]),decreasing=T))
#
# bDESIGN <- bDESIGN[!(bDESIGN$Label %in% samples_to_remove),]
# bDAT <- bDAT[,!(colnames(bDAT) %in% samples_to_remove)]
# ASVs_to_remove <- names(which(rowSums(bDAT) == 0))
# bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

print("No samples with low sequencing depth removed")

#this samples were not demultiplexed (reason unkown)
#remove from DESIGN
samples_to_remove <- bDESIGN$Label[!(bDESIGN$Label %in% colnames(bDAT))]
bDESIGN <- bDESIGN[!(bDESIGN$Label %in% samples_to_remove),]

print("samples removed (not sequenced)")
print(as.character(samples_to_remove))

#seq numbers
paste ("sum:", sum(colSums(bDAT)))
paste ("range:", range(colSums(bDAT)))
paste ("median:", median(colSums(bDAT)))

## define rarefication threshold 
#(number of reads from the sample with the lowest seq-depth)
b_rare <- min(colSums(bDAT))
b_rare <- 25000 #set to a round nbr

```

### Fungi

```{r seq numbers fungi, echo=F, message=F, warning=F, fig.align='left'}

### FUNGI ###

#show sorted seq numbers
#sort(colSums(fDAT),decreasing = T)

#remove samples with very low seq numbers
threshold <-9000
samples_to_remove <- names(which(colSums(fDAT) < threshold))
print("samples removed (low sequencing depth)")
pander(sort(colSums(fDAT[,colnames(fDAT) %in% samples_to_remove]),decreasing=T), caption="Fungi: seq depth of removed samples")

fDESIGN <- fDESIGN[!(fDESIGN$Label %in% samples_to_remove),]
fDAT <- fDAT[,!(colnames(fDAT) %in% samples_to_remove)]
ASVs_to_remove <- names(which(rowSums(fDAT) == 0))
fTAX <- fTAX[!(rownames(fTAX) %in% ASVs_to_remove),]

#this samples were not demultiplexed (reason unkown)
#remove from DESIGN
samples_to_remove <- fDESIGN$Label[!(fDESIGN$Label %in% colnames(fDAT))]
fDESIGN <- fDESIGN[!(fDESIGN$Label %in% samples_to_remove),]

print("samples removed (not sequenced)")
print(as.character(samples_to_remove))

#seq numbers
paste ("sum:", sum(colSums(fDAT)))
paste ("range:", range(colSums(fDAT)))
paste ("median:", median(colSums(fDAT)))

## define rarefaction threshold 
#(number of reads from the sample with the lowest seq-depth)
f_rare <- min(colSums(fDAT))
f_rare <- 9000 #set to a round nbr

```

We define the rarefaction threshold per sample to `r b_rare` for bacteria and `r f_rare` for fungi .

## Figure 1 | Sequencing depth

\vspace{5mm}

```{r Figure 1 bacteria, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

## boxplot
df <- data.frame(col_sum=colSums(bDAT),colnames=colnames(bDAT), Timepoint=bDESIGN$Timepoint, 
                 Arsenic=bDESIGN$Arsenic, group=bDESIGN$group, Experiment=bDESIGN$Experiment, Timepoint=as.factor(bDESIGN$Timepoint))

ylim2=boxplot.stats(df$col_sum)$stats[c(1, 5)]

seq_nr <- ggplot(data=df, aes(x=Timepoint, y=col_sum, fill=Arsenic)) + 
              geom_jitter(col="gray40") + 
              geom_boxplot(position=position_dodge2(width=0.75, preserve="single"),outlier.colour = NA) + 
              theme_bw() +
              theme(legend.position="none") +
              ylab("seq") +
              scale_fill_manual(values=level_cols_Arsenic) +
              #scale_color_manual(values=c("gray40","gray80")) +
              coord_cartesian(ylim=ylim2*1.05) +  
              theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
              #facet_grid(.~ Timepoint, scales='free_x', space="free")+
              ggtitle("Bacteria: Sequencing depth")

## barplot
df <- df[order(df$col_sum, decreasing = T),]
# df <- df[order(df$Timepoint),]
# temp <- table(df$Timepoint)
# x <- vector()
# for (i in 1:length(temp)){x <- c(x, 1:temp[i])}
# df$x <- x
df$x <- 1:length(df$col_sum)



samples <-  ggplot(df, aes(x=x, y=col_sum, fill=Arsenic, color=Arsenic))+
                geom_bar(aes(width=1), size=0.1, stat = "identity", col="black", lty=0.5)+
                #scale_y_continuous(trans = 'log2')+
                scale_fill_manual(values=level_cols_Arsenic) +
                scale_colour_manual(values = level_cols_Arsenic) +
                xlab("samples")+
                ylab("seq / samples")+
                theme_bw() +
                theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
                      strip.background = element_blank(), strip.text.x = element_blank(), legend.position = "none")
                #facet_grid(.~ Timepoint, scales='free_x', space="free")
                #ggtitle("Sequencing depth per sample")


#combine the two plots (library cowplot)
bac_plot <- plot_grid(seq_nr + theme(legend.position="none"),
               samples + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               rel_heights = c(9,5),
               nrow=2)

```

```{r Figure 1 fungi, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

## boxplot
df <- data.frame(col_sum=colSums(fDAT),colnames=colnames(fDAT), Timepoint=fDESIGN$Timepoint, 
                 Arsenic=fDESIGN$Arsenic, group=fDESIGN$group, Experiment=fDESIGN$Experiment, Timepoint=as.factor(fDESIGN$Timepoint))
#levels(df$Timepoint) <- c("day 0","day 2","day 5")

ylim2=boxplot.stats(df$col_sum)$stats[c(1, 5)]

seq_nr <- ggplot(data=df, aes(x=Timepoint, y=col_sum, fill=Arsenic)) + 
              geom_jitter(col="gray40") + 
              geom_boxplot(position=position_dodge2(width=0.75, preserve="single"),outlier.colour = NA) + 
              theme_bw() +
              theme(legend.position="none") +
              ylab("seq") +
              scale_fill_manual(values=level_cols_Arsenic) +
              #scale_color_manual(values=c("gray40","gray80")) +
              coord_cartesian(ylim=ylim2*1.05) +  
              theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
              #facet_grid(.~ Timepoint, scales='free_x', space="free")+
              ggtitle("Fungi: Sequencing depth")

## barplot
df <- df[order(df$col_sum, decreasing = T),]
#df <- df[order(df$Timepoint),]
# temp <- table(df$Timepoint)
# x <- vector()
# for (i in 1:length(temp)){x <- c(x, 1:temp[i])}
# df$x <- x
df$x <- 1:length(df$col_sum)


samples <-  ggplot(df, aes(x=x, y=col_sum, fill=Arsenic, color=Arsenic))+
                geom_bar(aes(width=1), size=0.1, stat = "identity", col="black", lty=0.5)+
                #scale_y_continuous(trans = 'log2')+
                scale_fill_manual(values=level_cols_Arsenic) +
                scale_colour_manual(values = level_cols_Arsenic) +
                xlab("samples")+
                ylab("seq / sample")+
                theme_bw() +
                theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
                      strip.background = element_blank(), strip.text.x = element_blank(), legend.position = "none")
                #facet_grid(.~ Timepoint, scales='free_x', space="free")
                #ggtitle("Sequencing depth per sample")


#combine the two plots (library cowplot)
fun_plot <- plot_grid(seq_nr + theme(legend.position="none"),
               samples + theme(legend.position="none") + labs(colour=""),
               #labels=c("A","B"),
               align = "hv",
               rel_heights = c(9,5),
               nrow=2)

```

```{r Figure 1 combined, fig.height=5, fig.width=8, echo=F, warning=F, message=F}

#combine the two plots (library cowplot)
fig1 <- plot_grid(bac_plot + theme(legend.position="none"),
               fun_plot + theme(legend.position="none") + labs(colour=""),
               labels=c("A","B"),
               align = "hv",
               nrow=1)

#add legend
legend <- get_legend(seq_nr + theme(legend.position="bottom") + labs(fill="Arsenic"))
fig1 <- plot_grid(fig1, legend, nrow=2, rel_heights=c(1, .1))

#print
fig1


```

\pagebreak

# Data normalization

## Asymptotic Kruskal-Wallis Test
We test for different sequencing depths between the groups.

### Bacteria

```{r seq numbers per group bacteria, echo=F, message=F, warning=F}

bDAT_krusk <- kruskal_test(colSums(bDAT) ~ bDESIGN$group)  # library(coin)
bDAT_krusk

```

### Fungi

```{r seq numbers per group fungi, echo=F, message=F, warning=F}

fDAT_krusk <- kruskal_test(colSums(fDAT) ~ fDESIGN$group)  # library(coin)
fDAT_krusk

```
**Conclusion:** Because we find significant differences in sequencing depths between the sample groups for bacteria, we will normalize the data by rarefication for diversity comparisions (see Weiss et al. (2017), Microbiome Journal). For consistency we rarefy for bacteria and fungi.


```{r DAT rare, warning=F, echo=F}

# rarefication with library(vegan)
bDAT_rare <- t(rrarefy(t(bDAT), b_rare))
bDAT_rare <- bDAT_rare[rowSums(bDAT_rare) > 0,]  # removal of rows with 0 values

fDAT_rare <- t(rrarefy(t(fDAT), f_rare))
fDAT_rare <- fDAT_rare[rowSums(fDAT_rare) > 0,]  # removal of rows with 0 values

```

## Outlier Detection

We use the method CLOUD developped by Montassier et al. 2018, which is a non.parametric detection test for outliers. We perform the test with Bray-Curtis distances from the rarefied data. We set the number of nearest neighbors to 60% of the samples size and chose an empirical outlier percentile of 5%. We remove all outliers from our data.

### Bacteria

```{r bac outlier detection, warning=F, echo=F}

#function to detect outliers
detect_outliers <- function(DESIGN, DAT_rare){

  #subset
  samples_sub <- DESIGN$Label
  DAT_sub <- DAT_rare[,colnames(DAT_rare) %in% samples_sub]
  bdist <- vegdist(t(DAT_sub), method="bray")
  
  #test
  k <- floor(ncol(DAT_sub) * 0.6) #number of nearest neighbors, here we chose 60% of sample size
  test <- piecewise_kn_V1(d = bdist, test.ix = colnames(DAT_sub), k = k) #CLOUD test from Montassier et al. 2018
  outliers <- colnames(DAT_sub)[test[["pvals"]]<0.05] #outliers (p < 0.0.5)
  
  return(outliers)

}

### BACTERIA 
boutliers <- detect_outliers(DESIGN=bDESIGN, DAT_rare=bDAT_rare)

bDESIGN <- bDESIGN[!(bDESIGN$Label %in% boutliers),]
bDAT <- bDAT[,!(colnames(bDAT) %in% boutliers)]
bDAT_rare <- bDAT_rare[,!(colnames(bDAT_rare) %in% boutliers)]
ASVs_to_remove <- names(which(rowSums(bDAT_rare) == 0))
bTAX <- bTAX[!(rownames(bTAX) %in% ASVs_to_remove),]

print("samples removed (outliers)")
print(as.character(boutliers))

```

### Fungi

```{r fun outlier detection, warning=F, echo=F}

### FUNGI
foutliers <- detect_outliers(DESIGN=fDESIGN, DAT_rare=fDAT_rare)

fDESIGN <- fDESIGN[!(fDESIGN$Label %in% foutliers),]
fDAT <- fDAT[,!(colnames(fDAT) %in% foutliers)]
fDAT_rare <- fDAT_rare[,!(colnames(fDAT_rare) %in% foutliers)]
ASVs_to_remove <- names(which(rowSums(fDAT_rare) == 0))
fTAX <- fTAX[!(rownames(fTAX) %in% ASVs_to_remove),]

print("samples removed (outliers)")
print(as.character(foutliers))

```

\pagebreak

## Sample Size

We end up with the following number of samples per treatment for the analysis:

```{r sample size rare, warning=F, echo=F}

#final number of samples
btable <- table(bDESIGN$Timepoint,bDESIGN$Arsenic)
pander(btable, caption="Bacteria: Sample profiles")

ftable <- table(fDESIGN$Timepoint,fDESIGN$Arsenic)
pander(ftable, caption="Fungi: Sample profiles")

```

```{r export RDA files, echo=F, warning=F, message=F}

# Export

# create directory
dir.create("interim")
dir.create("interim/01_Import_Normalization")

## set output directory
setwd("interim/01_Import_Normalization")

#save objects needed in the following scripts
saveRDS(bDESIGN, "bDESIGN.RDS")
saveRDS(bDAT, "bDAT.RDS")
saveRDS(bTAX, "bTAX.RDS")
saveRDS(bDAT_rare, "bDAT_rare.RDS")
saveRDS(b_rare, "b_rare.RDS")

saveRDS(fDESIGN, "fDESIGN.RDS")
saveRDS(fDAT, "fDAT.RDS")
saveRDS(fTAX, "fTAX.RDS")
saveRDS(fDAT_rare, "fDAT_rare.RDS")
saveRDS(f_rare, "f_rare.RDS")

saveRDS(level_cols_Arsenic, "level_cols_Arsenic.RDS")
saveRDS(level_cols_Timepoint, "level_cols_Timepoint.RDS")
saveRDS(bTAX_level_cols_phyla, "bTAX_level_cols_phyla.RDS")
saveRDS(fTAX_level_cols_phyla, "fTAX_level_cols_phyla.RDS")

```

\pagebreak
